policy.go

```diff
@@ -399,218 +393,252 @@ func createPcrPolicyCounter(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKeyN
+// an adversary without the private part of the ephemeral key to undefine and redefine an identical NV index in order to remove the
+// TPMA_NV_READLOCKED attribute and reenable access to sealed keys. This was the approach previously taken by secboot, but this had
+// its own problem - if the NV index is accidentally undefined, then this would make all sealed keys permanently unrecoverable.
+// Ideally, provisioning the TPM without clearing it should be able to remedy any accidental changes and restore access to valid
+// sealed keys in all cases except where the storage primary seed has been changed (ie, the TPM has been cleared).
+//
+// The current approach used for protecting the NV index used by LockAccessToSealedKeys makes use of 2 NV indices and has the property
+// that the NV indices have names that are common to all devices and can always be recreated, but they have authorization policies
+// that enforce a creation sequence that means that they can only be created in the locked state. The way this works is as follows:
+// - A bootstrap index is created at handle 2 with TPMA_NV_AUTHREAD and TPMA_NV_AUTHWRITE attributes and an empty authorization
+//   policy.
+// - An empty write is performed to initialize the bootstrap index.
+// - An index is created at handle 1 with TPMA_NV_AUTHREAD, TPMA_NV_POLICY_WRITE and TPMA_NV_READ_STCLEAR attributes, and an
+//   authorization policy that can only be satisfied by the presence of the initialized bootstrap index.
+// - An empty write is performed to initialize index 1 using a policy session.
+// - The bootstrap index is undefined and another index is created at handle 2 with TPMA_NV_AUTHREAD and TPMA_NV_POLICY_WRITE
+//   attributes, and an authorization policy that can only be satisfied by the presence of the initialized index at handle 1 with
+//   its read lock enabled.
+// - The read lock for the index at handle 1 is enabled.
+// - An empty write is performed to initialize index 2 using a policy session.
+//
+// Authorization policies for sealed keys must assert the presence of both of these NV indices using the names of the initialized
+// indices without the read lock enabled. Enabling the read lock on index 1 by calling LockAccessToSealedKeys will disable access to
+// these keys.
+//
+// It isn't possible to recreate these indices in their unlocked state. Although they can be recovered if they are undefined
+// accidentally, they will not return to their unlocked state until the next TPM reset or restart.
+//
+// An adversary can't undefine and redefine either index in order to reenable access to sealed keys. Eg, say an adversary performs the
+// following actions:
+// - They undefine and redefine index 1. The new index isn't read locked, but sealed keys can't be unsealed until the new index has
+//   been written to. But it has an authorization policy that requires index 2 to be undefined and replaced with the bootstrap index.
```

Comment:
how does the binding of policy to indexes work?  from reading this sentence it sounds like it works in such a way that policy survives undefine/redefine of the index?

Reply:
I'm not sure I understand the question, but if you're referring to the policy of sealed keys, then they are bound to the indices using the names of them, which are dependent on the indices own authorization policy and public attributes. If these indices are undefined, it's possible to recover and initialize them with the same authorization policy as before, but they won't have attributes (and therefore, names) that permit unsealing of keys until after a restart.

Reply:
thx